models/ReadActivity.js

```diff
@@ -0,0 +1,149 @@
+ */
+class ReadActivity extends BaseModel {
+  static get tableName () /*: string */ {
+    return 'readActivity'
+  }
+
+  get path () /*: string */ {
+    return 'readActivity'
+  }
+
+  static get jsonSchema () /*: any */ {
+    return {
+      type: 'object',
+      properties: {
+        id: { type: 'string' },
+        selector: { type: 'object' },
+        json: { type: 'object' },
+        readerId: { type: 'string' },
+        publicationIn: { type: 'string' },
+        published: { type: 'string', format: 'date-time' }
+      },
+      required: ['readerId', 'publicationId']
+    }
+  }
+
+  static get relationMappings () /*: any */ {
+    const { Publication } = require('./Publication.js')
+    const { Reader } = require('./Reader.js')
+    const { Document } = require('./Document.js')
+    const { Note } = require('./Note.js')
+    const { Tag } = require('./Tag.js')
+
+    /*
+    assert.ok(Model)
+    assert.ok(Model.BelongsToOneRelation)
+    assert.ok(Reader)
+    assert.ok(Document)
+    assert.ok(Note)
+    assert.ok(Publication)
+    */
+    return {
+      reader: {
+        relation: Model.BelongsToOneRelation,
+        modelClass: Reader,
+        join: {
+          from: 'readActivity.readerId',
+          to: 'Reader.id'
+        }
+      },
+      publication: {
+        relation: Model.BelongsToOneRelation,
+        modelClass: Publication,
+        join: {
+          from: 'readActivity.publicationId',
+          to: 'Publication.id'
+        }
+      }
+    }
+  }
+
+  // static async byId (id /*: string */) /*: Promise<activity> */ {
+  //   return await ReaderActivity.query()
+  //     .findById(id)
+  //     .eager('readerActivity')
+  // }
+
+  static async createReadActivity (
+    readerId /*: string */,
+    publicationId /*: string */,
+    object /*: any */
+  ) /*: any */ {
+    if (!readerId) return new Error('wrong readerId')
+
+    if (!publicationId) return new Error('wrong publicationId')
+
+    if (!object) return new Error('wrong object')
+
+    const props = _.pick(object, ['selector', 'json'])
+
+    props.id = createId()
+    props.readerId = readerId
+    props.publicationId = publicationId
+    props.published = new Date().toISOString()
+
+    try {
+      var activity = await ReadActivity.query()
+        .insert(props)
+        .returning('*')
+
+      console.log('LAST ACTIVITY INSERTED: ' + activity.id)
+
+      return activity
+    } catch (err) {
+      if (err.constraint === 'readactivity_readerid_foreign') {
+        return new Error('no reader')
+      } else if (err.constraint === 'readactivity_publicationid_foreign') {
+        return new Error('no publication')
+      }
+    }
+  }
+
+  static async getLatestReadActivity (publicationId /*: string */) /*: any */ {
+    if (!publicationId) return new Error('wrong publicationId')
+
+    const readActivities = await ReadActivity.query()
+      .where('publicationId', '=', publicationId)
+      .orderBy('published')
```

Comment:
can you do it with .max('published') instead? There might be a lot of read activities after a while, and there is no point in fetching all of them if we only need one.